airsonic-main/src/main/java/org/airsonic/player/service/search/IndexManager.java

```diff
@@ -1,169 +1,236 @@
+
+    private Map<IndexType, SearcherManager> searchers = new ConcurrentHashMap<>();
+
+    private Map<IndexType, IndexWriter> writers = new ConcurrentHashMap<>();
+
+    public void index(Album album) {
+        Term primarykey = documentFactory.createPrimarykey(album);
+        Document document = documentFactory.createAlbumId3Document(album);
+        try {
+            writers.get(IndexType.ALBUM_ID3).updateDocument(primarykey, document);
+        } catch (Exception x) {
+            LOG.error("Failed to create search index for " + album, x);
+        }
+    }
+
+    public void index(Artist artist, MusicFolder musicFolder) {
+        Term primarykey = documentFactory.createPrimarykey(artist);
+        Document document = documentFactory.createArtistId3Document(artist, musicFolder);
+        try {
+            writers.get(IndexType.ARTIST_ID3).updateDocument(primarykey, document);
+        } catch (Exception x) {
+            LOG.error("Failed to create search index for " + artist, x);
+        }
+    }
+
+    public void index(MediaFile mediaFile) {
+        Term primarykey = documentFactory.createPrimarykey(mediaFile);
+        try {
+            if (mediaFile.isFile()) {
+                Document document = documentFactory.createSongDocument(mediaFile);
+                writers.get(IndexType.SONG).updateDocument(primarykey, document);
+            } else if (mediaFile.isAlbum()) {
+                Document document = documentFactory.createAlbumDocument(mediaFile);
+                writers.get(IndexType.ALBUM).updateDocument(primarykey, document);
+            } else {
+                Document document = documentFactory.createArtistDocument(mediaFile);
+                writers.get(IndexType.ARTIST).updateDocument(primarykey, document);
+            }
+        } catch (Exception x) {
+            LOG.error("Failed to create search index for " + mediaFile, x);
+        }
+    }
+
+    private String getIndexRootDirectoryName() {
+        return INDEX_DIR;
+    }
+
+    public final void startIndexing() {
+        try {
+            for (IndexType IndexType : IndexType.values()) {
+                writers.put(IndexType, createIndexWriter(IndexType));
+            }
+        } catch (IOException e) {
+            LOG.error("Failed to create search index.", e);
+        }
+    }
+
+    /**
+     * 
+     * @param indexType
+     * @return
+     * @throws IOException
+     */
+    private IndexWriter createIndexWriter(IndexType indexType) throws IOException {
+        File dir = util.getDirectory.apply(getIndexRootDirectoryName(), indexType);
+        IndexWriterConfig config = new IndexWriterConfig(analyzerFactory.getAnalyzer());
+        return new IndexWriter(FSDirectory.open(dir.toPath()), config);
+    }
+
+    /**
+     * Close Writer of all indexes and update SearcherManager.
+     * Called at the end of the Scan flow.
+     */
+    public void stopIndexing() {
+        Arrays.asList(IndexType.values()).forEach(i -> stopIndexing(i));
```

Comment:
Can use method reference here. I.e. replace `i -> stopIndexing(i)` with `this::stopIndexing`.

Reply:
I revised that according to your advice.